### PR TITLE
Schedule Activities Based On External Events

### PR DESCRIPTION
**Pull request by pranav-super**
* **Tickets addressed:** MPS-124
* **Review:** By commit
* **Merge strategy:** Merge (no squash)

## Description
Following the introduction of External Events in [#1513](https://github.com/NASA-AMMOS/aerie/pull/1513), this PR introduces the ability to schedule Activities based off of External Event occurrences, using the procedural scheduler.

This PR introduces 4 new types which users can use in their definition of scheduling procedures to narrow down the external events they would like to schedule off of and which properties they would like to access:
- an `ExternalEvent` type, which contains event-specific properties,
- an `ExternalEvents` type, which represents a filterable collection of said events,
- an `EventQuery` type, which can be used to filter the set of all `ExternalEvent`s associated with a given plan by their `ExternalSource`, derivation group, or event type.
- and an `ExternalSource` type, which can be used by `EventQuery`s to narrow down the set of events being examined by a procedure to those contained in specific sources.

To make this more explicitly clear, we will illustrate an example,  based on the tests in `ExternalEventsTests.java` under the `e2e-tests` directory.

Consider this set of events:
image

Consider scheduling the Banananation activity `BiteBanana` off of these events. 
This can mean a variety of different things, depending on the exact behavior we desire:
- schedule the activity coincident with all events
- schedule the activity coincident with events belonging to derivation group `"TestGroup"`
- schedule the activity coincident with events belonging to the second source
- schedule the activity based on events of type `"Test_Type"`
- schedule the activity based on events with the substring `"01"` in their key.

For the sake of brevity, we will explore just one case - events belonging to the second source with the substring `"01"` in their key. This case makes use of all of the types described above.

```
@SchedulingProcedure
public record ExternalEventsSourceQueryGoal() implements Goal {
  @Override
  public void run(@NotNull final EditablePlan plan) {

    // extract events belonging to the second source
    EventQuery eventQuery = new EventQuery(
        null,
        null,
        List.of(new ExternalSource("NewTest.json", "TestGroup_2"))
    );

    for (final var e: plan.events(eventQuery)) {
      // filter events that we schedule off of by key
      if (e.key.contains("01")) {
        plan.create(
            "BiteBanana",
            // place the directive such that it is coincident with the event's start
            new DirectiveStart.Absolute(e.getInterval().start),
            Map.of("biteSize", SerializedValue.of(1)));
      }
    }
    plan.commit();
  }
}
```

This goal does the following:
1. create an `EventQuery` to select events belonging to the second source. To specify that selector, we create an `ExternalSource` object, whose properties are the two properties necessary to uniquely identify two `ExternalSource`s in AERIE: a `key` and a `derivation_group_name`. 
2. Call `plan.events(eventQuery)`. This returns all external events associated with a plan (based on the associated derivation groups), and filters them according to the passed-in query.
3. Filter by key. We directly access the properties of each event (namely, `key`) now to further match our requirements.
4. Create the activity, placing it coincident with the event's start, accessed via `getInterval().start`.

After running it, we get the following result in AERIE:
image

As expected, there is a single activity, scheduled off of an `Event_01` from the second source.

## Verification
The changes were verified both by testing with our own models, and subsequent testing of the 4 new types listed above in a few end-to-end tests (see `ExternalEventsTests.java` under the `e2e-tests` directory).

## Documentation
Documentation will need to be added to [aerie-docs](https://vcs.ssmo.gsfc.nasa.gov/SSMO-MPS/aerie-docs/tree/master/docs) to describe how scheduling using external events work. As this feature is solely an addition, no existing documentation is invalidated; the documentation is simply expanded.

## Future work
The next steps for External Event work are:
- additional cleanup and optimization of External Events in the front and backend
- adding properties to events and sources, and exposing those to the scheduler
- integration with the procedural constraint system
- introduction of a formal JSON schema for events and sources and support of said schema in the database

### CODE REVIEW

**Comment by JoelCourtney** [comment]
@pranav-super you forgot to push the change to GraphQLMerlinDatabaseService

File: scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java [hunk 1]
```diff
@@ -609,7 +617,7 @@ private Optional<MissingActivityNetworkConflict> makeActivityNetworkConflict(
         curPreviouslyInstantiated = null;
       }
       taskNetwork.startsAfterStart(
-          allActivitiesInNetwork.getLast(),
```
Comment (JoelCourtney) [inline scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java]: why did you change this?

Comment (pranav-super) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java]: IntelliJ was getting very upset with me using `getLast()` and `getFirst()`, saying it wasn't supported on `List`. Checking the `List` interface supported this assertion, as did [this](https://stackoverflow.com/questions/1049518/why-does-the-java-list-interface-not-support-getlast). I'm pretty skeptical but this was the only way I could compile and get things to run!

Comment (JoelCourtney) [reply scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java]: It is supported in Java 21, so I think the problem must be in your setup, not the code. Please revert this.

On Mac in IntelliJ, you can go to `File > Project Structure > Project > SDK` to change your jdk version.

**Comment by JoelCourtney** [comment]
Rebased to clean up the git history

File: procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt [hunk 2]
```diff
@@ -41,4 +42,13 @@ interface Plan {
    * @param name string name of the resource
    */
   fun <V: Any, TL: SerialSegmentOps<V, TL>> resource(name: String, deserializer: (List<Segment<SerializedValue>>) -> TL): TL
+
+  /** Get external events associated with this plan. */
+  fun events(query: EventQuery): ExternalEvents
```
Comment (skovati) [inline procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt:47]: I'm not quite seeing the purpose of `EventQuery`, since we already provide filtering capabilities in these other event function overloads. Why not just have a function in the `Plan` interface here with nullable arguments (`derivationGroup: String?, eventType: String?, source: String?`) that conditionally handles the filtering?

Not a big deal to me, but it does seem a little confusing that we're exposing multiple filtering APIs to users, especially when they could do this filtering in "userspace" anyways.

File: procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/collections/ExternalEvents.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+data class ExternalEvents(private val timeline: Timeline<ExternalEvent, ExternalEvents>):
+    Timeline<ExternalEvent, ExternalEvents> by timeline,
+    NonZeroDurationOps<ExternalEvent, ExternalEvents>,
+    ParallelOps<ExternalEvent, ExternalEvents>
+{
+  constructor(vararg events: ExternalEvent): this(events.asList())
+  constructor(events: List<ExternalEvent>): this(BaseTimeline(::ExternalEvents, preprocessList(events, null)))
+
+  /** Filter by one or more types. */
+  fun filterByType(vararg types: String) = filter { it.type in types }
```
Comment (skovati) [inline procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/collections/ExternalEvents.kt:24]: This looks like yet another filtering API we're exposing to users. When would we expect users to use these functions v.s. querying while getting the `ExternalEvents` object?

**Review by skovati — CHANGES_REQUESTED** [review]
Looks great so far! Nice work. Just have a few questions about streamlining the filtering API we're exposing to users to simplify things.

File: procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt (hunk 2 above)
Comment (pranav-super) [reply procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt:47]: The only thing this provides in addition is the ability to specify multiple filters in one single query, without needing multiple filtering function calls. They are be functionally equivalent, but, at least in testing of the feature, this query did seem to be pretty helpful. 

@JoelCourtney Any thoughts on this? Should we remove either the `EventQuery` or the other function options?

File: procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/collections/ExternalEvents.kt (hunk 3 above)
Comment (pranav-super) [reply procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/collections/ExternalEvents.kt:24]: If there are conditional filters we want to apply on an `ExternalEvents` object, this is how we could do it without having to pull repeatedly from the `Plan` object.

File: procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt (hunk 2 above)
Comment (skovati) [reply procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt:47]: Gotcha, I just worry about exposing too many filtering objects to end users. It looks like we can filter when requesting events in two ways:
1. `val events = plan.events("SomeGroupName")`
2. `val events = plan.events(new EventQuery("SomeGroupName"))`

And once we have an `ExternalEvents` object, we can filter in two ways:
1. `val justFooType = events.filterByType("Foo")`
2. `val justFooType = events.filter { it.type == "Foo }`

Like you said, all are functionally equivalent, but I can imagine a user being a bit confused when they see multiple ways to filter events without realizing they're functionally equivalent.

Now (while introducing this feature) is a good time to think about the most clear / concise API we can expose, since future changes would (likely?) be breaking.

Comment (pranav-super) [reply procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt:47]: Sure! I think allowing filtering from the `Plan` and filtering from the `ExternalEvents` objects should both be available, but I agree we should limit each to only 1 method.

Comment (pranav-super) [reply procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt:47]: Actually wait - I think I change my mind. I think we should actually allow all of these. The point of the methods on `Plan`, despite the fact that they can be replicated in `EventQuery`, is to provide idiomatic ways of filtering events from a plan without needing to specify `EventQuery`. We will add more to events (like metadata) in the future and as such having both an easy idiomatic way as well as a more complicated, query way, to fetch events I would argue is helpful. At the current moment there is complete overlap but going forward I don't think this will be the case!

Comment (skovati) [reply procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt:47]: Gotcha! That's why I had originally asked about the point of `EventQuery` since it (currently) feels like an unnecessary layer of indirection around simple filtering logic, but if you see the `EventQuery` object encapsulating logic around more complex filters (e.g. some sort of metadata many-to-one mapping or something), then I'm fine keeping it in.

File: e2e-tests/src/main/java/gov/nasa/jpl/aerie/e2e/procedural/scheduling/procedures/ExternalEventsSourceQueryGoal.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+@SchedulingProcedure
+public record ExternalEventsSourceQueryGoal() implements Goal {
+  @Override
+  public void run(@NotNull final EditablePlan plan) {
+
+    // extract events belonging to the second source
+    EventQuery eventQuery = new EventQuery(
+        null,
+        null,
+        List.of(new ExternalSource("NewTest.json", "TestGroup_2"))
```
Comment (skovati) [inline e2e-tests/src/main/java/gov/nasa/jpl/aerie/e2e/procedural/scheduling/procedures/ExternalEventsSourceQueryGoal.java:24]: Is it safe to assume that users will know external event source file names _a priori_? Or would it be helpful to include some way to query for which sources had been registered with a plan?

I'm thinking of cases where source file names might be autogenerated or have some unknown timestamp, and users might just want to reference events where `source.includes("DSN")` or something.

While typing this out, I'm realizing that particular use case could be accomplished by "manually" filtering on event sources after we've run the `EventQuery` against the `Plan` 🤔

Comment (pranav-super) [reply e2e-tests/src/main/java/gov/nasa/jpl/aerie/e2e/procedural/scheduling/procedures/ExternalEventsSourceQueryGoal.java:24]: Not necessarily a safe assumption; as such it is possible to allow users to filter for that kind of thing: 
```
var events = plan.events();
events.filter(false, event -> event.source.key.contains("DSN"));
```

Comment (skovati) [reply e2e-tests/src/main/java/gov/nasa/jpl/aerie/e2e/procedural/scheduling/procedures/ExternalEventsSourceQueryGoal.java:24]: My thoughts exactly!

**Review by skovati — APPROVED** [review]

File: procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt (hunk 2 above)
Comment (JoelCourtney) [reply procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt:47]: Actually EventQuery and the other `events` overloads could provide some extra utility, and the explicit filter functions are just sugar. I made EventQuery to give the Plan implementation the option to optimize the events database lookup so that it only loads the events you need (although the built in scheduler doesn't actually do this).

I think there are situations that are best suited to each of the three ways of filtering, but I agree that having three ways is kinda overkill. So I'd be happy removing the `events(derivGroup)` and `events(group, type)` overloads, since I think they're the least useful.

Comment (pranav-super) [reply procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/plan/Plan.kt:47]: These are the ones defined as methods on `Plan`, right?
EDIT: they have been factored out.
